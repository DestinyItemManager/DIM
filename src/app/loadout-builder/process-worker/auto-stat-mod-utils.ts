import { compareBy } from 'app/utils/comparators';
import { ArmorStatHashes, artificeStatBoost, majorStatBoost, minorStatBoost } from '../types';
import { LoSessionInfo } from './process-utils';
import { AutoModData } from './types';

/**
 * A particular way of achieving a target stat value (for a single stat).
 */
export interface ModsPick {
  /** The number of artifice mods this pick contains. */
  numArtificeMods: number;
  /** The number of general mods this pick contains. */
  numGeneralMods: number;
  /** The cost of the general mods this pick contains, sorted descending. */
  generalModsCosts: number[];
  /** General + artifice mod hashes */
  modHashes: number[];
  /** Sum of generalModCosts */
  modEnergyCost: number;
  /** Which stat this set of mods targets */
  targetStatIndex: number;
  /** The exact number of points this set of mods provides (if we ask for 1 stat point, an artifice mod might give 3) */
  exactStatPoints: number;
}

/**
 * Precalculated ways of hitting all possible stat values for a single stat.
 */
interface CacheForStat {
  statMap: {
    [targetStat: number]: ModsPick[] | undefined;
  };
}

/**
 * Precalculated ways of hitting stat values, separated by stat hash.
 */
export interface AutoModsMap {
  statCaches: { [targetStatIndex: number]: CacheForStat };
  /**
   * See comments in pickOptimalStatMods. That function optimizes for total tier first,
   * so if a less-prioritized stat also has more costly mods, then it cannot result in a higher
   * total tier.
   * So for each stat index, this contains a list of stats where the stat mods are better
   * for purposes of optimizing total tier, by having cheaper or more mods available.
   */
  cheaperStatRelations: { [statIndex: number]: ArmorStatHashes[] };
}

/**
 * Pick auto mods (general mods and artifice mods)
 * that satisfy the given `neededStats` in `statOrder`.
 */
export function chooseAutoMods(
  info: LoSessionInfo,
  neededStats: number[],
  numArtificeMods: number,
  remainingEnergyCapacities: number[][],
  remainingTotalEnergy: number,
) {
  return recursivelyChooseMods(
    info,
    neededStats,
    0,
    info.numAvailableGeneralMods,
    numArtificeMods,
    remainingEnergyCapacities,
    remainingTotalEnergy,
    undefined,
  );
}

function doGeneralModsFit(
  info: LoSessionInfo,
  /** variants of remaining energy capacities given our activity mod assignment, each sorted descending */
  remainingEnergyCapacities: number[][],
  pickedMods: ModsPick[] | undefined,
) {
  // These are already sorted
  let generalModCosts = info.generalModCosts;
  // eslint-disable-next-line @typescript-eslint/prefer-optional-chain
  if (pickedMods !== undefined && pickedMods.length) {
    generalModCosts = generalModCosts.slice();
    // Intentionally open-coded for performance
    // eslint-disable-next-line @typescript-eslint/prefer-for-of
    for (let i = 0; i < pickedMods.length; i++) {
      generalModCosts.push(...pickedMods[i].generalModsCosts);
    }
    generalModCosts.sort((a, b) => b - a);
  }

  return remainingEnergyCapacities.some((capacities) =>
    generalModCosts.every((cost, index) => cost <= capacities[index]),
  );
}

/**
 * Find a combination of artifice and general mods that can
 * help hit the `neededStats` starting from `statIndex` by recursively
 * enumerating all combinations.
 * `pickedMods` contains the mods chosen for earlier stats.
 */
function recursivelyChooseMods(
  info: LoSessionInfo,
  neededStats: number[],
  statIndex: number,
  remainingGeneralSlots: number,
  remainingArtificeSlots: number,
  /** variants of remaining energy capacities given our activity mod assignment, each sorted descending */
  remainingEnergyCapacities: number[][],
  remainingTotalEnergy: number,
  pickedMods: ModsPick[] | undefined,
): ModsPick[] | undefined {
  while (statIndex < neededStats.length && neededStats[statIndex] === 0) {
    statIndex++;
  }

  if (statIndex === neededStats.length) {
    // We've hit the end of our needed stats, check if this is possible
    if (doGeneralModsFit(info, remainingEnergyCapacities, pickedMods)) {
      return pickedMods ?? [];
    } else {
      return undefined;
    }
  }

  const possiblePicks = info.autoModOptions.statCaches[statIndex].statMap[neededStats[statIndex]];
  if (!possiblePicks) {
    // we can't possibly hit our target stats
    return undefined;
  }

  // Create a new array we append the pick for this stat to.
  const subArray = pickedMods !== undefined ? pickedMods.slice() : [];
  // Dummy value just so we don't repeatedly push and pop.
  subArray.push(subArray[0]);

  for (const pick of possiblePicks) {
    if (
      pick.numArtificeMods > remainingArtificeSlots ||
      pick.numGeneralMods > remainingGeneralSlots ||
      pick.modEnergyCost > remainingTotalEnergy
    ) {
      continue;
    }
    subArray[subArray.length - 1] = pick;
    const solution = recursivelyChooseMods(
      info,
      neededStats,
      statIndex + 1,
      remainingGeneralSlots - pick.numGeneralMods,
      remainingArtificeSlots - pick.numArtificeMods,
      remainingEnergyCapacities,
      remainingTotalEnergy - pick.modEnergyCost,
      subArray,
    );
    if (solution) {
      return solution;
    }
  }
}

/**
 * Previously we could use a simple algorithm to come up with all mod combinations to hit certain target stats
 * based on simple "mod splitting": Since minor and major mods give 5 and 10 respectively (common divisor 5),
 * we only needed to care about stat multiples of 5. And we could just start with +10 mods and derive variants by
 * splitting +10 mods into +5 mods. With stats ranging from 0 to 50 (11 values) and 6 stats,
 * this would end up with about 700,000 combinations that could be computed and cached on-demand.
 *
 * However, now we need to take care of the +3 artifice mods, and 3 is coprime with 5 and 10. So first of all,
 * it's a lot more difficult to come up with the pick variants that could hit certain stats, and even if we did,
 * we cannot efficiently cache the results since every single point matters and cache entries for 50^6 values would
 * mean our cache would simply explode. So instead we separate the caches by stat and then piece together the mod
 * picks when actually looking at sets.
 *
 * This unfortunately means a lot of `flatMap`ing down the road and is a lot less efficient. Improvements here
 * could make things a bit faster, especially when they remove equivalent combinations.
 */
function buildCacheForStat(
  autoModOptions: AutoModData,
  statHash: ArmorStatHashes,
  statIndex: number,
  availableGeneralStatMods: number,
) {
  const cache: CacheForStat = { statMap: {} };
  // Note: All of these could be undefined for whatever reason.
  // In that case, the loop bounds are 0 <= numMods <= 0.
  // Major and minor mod always exist together or not at all.
  const artificeMod = autoModOptions.artificeMods[statHash];
  const minorMod = autoModOptions.generalMods[statHash]?.minorMod;
  const majorMod = autoModOptions.generalMods[statHash]?.majorMod;

  for (let numArtificeMods = 0; numArtificeMods <= (artificeMod ? 5 : 0); numArtificeMods++) {
    for (
      let numMinorMods = 0;
      numMinorMods <= (minorMod ? availableGeneralStatMods : 0);
      numMinorMods++
    ) {
      for (
        let numMajorMods = 0;
        numMajorMods <= (majorMod ? availableGeneralStatMods - numMinorMods : 0);
        numMajorMods++
      ) {
        const statValue =
          numArtificeMods * artificeStatBoost +
          numMinorMods * minorStatBoost +
          numMajorMods * majorStatBoost;
        if (statValue === 0) {
          continue;
        }
        // We are allowed to provide more stat points than needed -- within reason.
        // If we have a major mod, this satisfies stat needs of 6,7,8,9,10
        // 5 can be satisfied by a strictly better pick that includes only a minor mod
        // If we have a major mod and an artifice mod, this satisfies 11,12,13.
        // 10 can be satisfied by dropping the artifice mod.
        // So if we have any artifice pieces, we are allowed to overshoot by 2, and if
        // not then we're allowed to overshoot by 4.
        // This ensures pareto-optimality of the various ways of hitting a stat target.
        // Note: Assumptions here are artificeStatBoost < minorStatBoost
        // and majorStatBoost = 2 * minorStatBoost
        const lowerRange =
          statValue - (numArtificeMods > 0 ? artificeStatBoost - 1 : minorStatBoost - 1);
        const obj: ModsPick = {
          numArtificeMods,
          numGeneralMods: numMinorMods + numMajorMods,
          generalModsCosts: [
            ...Array<number>(numMajorMods).fill(majorMod?.cost ?? 0),
            ...Array<number>(numMinorMods).fill(minorMod?.cost ?? 0),
          ],
          modHashes: [
            ...Array<number>(numMajorMods).fill(majorMod?.hash ?? 0),
            ...Array<number>(numMinorMods).fill(minorMod?.hash ?? 0),
            ...Array<number>(numArtificeMods).fill(artificeMod?.hash ?? 0),
          ],
          modEnergyCost:
            numMinorMods * (minorMod?.cost ?? 0) + numMajorMods * (majorMod?.cost ?? 0),
          targetStatIndex: statIndex,
          exactStatPoints: statValue,
        };
        for (let achievableValue = lowerRange; achievableValue <= statValue; achievableValue++) {
          (cache.statMap[achievableValue] ??= []).push(obj);
        }
      }
    }
  }

  for (const pickArray of Object.values(cache.statMap)) {
    pickArray!.sort(compareBy((pick) => -pick.numArtificeMods));
  }

  return cache;
}

/**
 * See comments in pickOptimalStatMods. That function optimizes for total tier first,
 * so if a less-prioritized stat also has more costly mods, then it cannot result in a higher
 * total tier.
 * So for each ArmorStatHash, this builds a list of stats where the stat mods are better
 * for purposes of optimizing total tier, by having cheaper or more mods available.
 */
function buildLessCostlyRelations(
  autoModOptions: AutoModData,
  availableGeneralStatMods: number,
  statOrder: ArmorStatHashes[],
) {
  return Object.fromEntries(
    statOrder.map((armorStat1, statIndex1) => {
      const betterStatIndices: number[] = [];
      for (const [statIndex2, armorStat2] of statOrder.entries()) {
        if (availableGeneralStatMods === 0) {
          // No general mods means it doesn't matter how much our general mods actually cost
          if (!autoModOptions.artificeMods[armorStat1] || autoModOptions.artificeMods[armorStat2]) {
            // So if Stat1 has no artifice mods, or Stat2 has them, Stat2 can do equal or better
            betterStatIndices.push(statIndex2);
          }
        } else {
          const mods1 = autoModOptions.generalMods[armorStat1];
          const mods2 = autoModOptions.generalMods[armorStat2];

          if (autoModOptions.artificeMods[armorStat1] && !autoModOptions.artificeMods[armorStat2]) {
            // Stat1 has artifice mods, Stat2 doesn't, so Stat2 is worse in that aspect
          } else if (!mods1) {
            // Stat1 has no mods, so Stat2 can always do equal or better
            betterStatIndices.push(statIndex2);
          } else if (!mods2) {
            // Stat1 has mods, Stat2 doesn't, so Stat2 is worse in that aspect
          } else {
            const [large1Cost, large2Cost] = [mods1.majorMod.cost, mods2.majorMod.cost];
            const [small1Cost, small2Cost] = [mods1.minorMod.cost, mods2.minorMod.cost];
            // mods for armorStat2 are cheaper (dominate armorStat1) if
            // they're cheaper or same
            if (small1Cost >= small2Cost && large1Cost >= large2Cost) {
              betterStatIndices.push(statIndex2);
            }
          }
        }
      }

      return [statIndex1, betterStatIndices];
    }),
  ) as AutoModsMap['cheaperStatRelations'];
}

export function buildAutoModsMap(
  autoModOptions: AutoModData,
  availableGeneralStatMods: number,
  statOrder: ArmorStatHashes[],
): AutoModsMap {
  return {
    statCaches: Object.fromEntries(
      statOrder.map((statHash, statIndex) => [
        statIndex,
        buildCacheForStat(autoModOptions, statHash, statIndex, availableGeneralStatMods),
      ]),
    ) as AutoModsMap['statCaches'],
    cheaperStatRelations: buildLessCostlyRelations(
      autoModOptions,
      availableGeneralStatMods,
      statOrder,
    ),
  };
}
