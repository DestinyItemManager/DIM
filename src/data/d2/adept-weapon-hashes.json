[
  276080079,
  532746994,
  1173780905,
  1366917989,
  1481892490,
  2147010335,
  2386979999,
  2527666306,
  3514144928,
  3637570176,
  3836861464,
  3847137620
]
