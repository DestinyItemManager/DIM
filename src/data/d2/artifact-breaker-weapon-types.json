{
  "485622768": [
    14,
    10
  ],
  "2611060930": [
    12,
    2489664120
  ],
  "3178805705": [
    9,
    3871742104,
    13
  ]
}
