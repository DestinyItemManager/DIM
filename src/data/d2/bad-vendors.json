[2677528157, 3231773039]
