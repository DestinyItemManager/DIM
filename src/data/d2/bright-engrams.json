{
  "1": 3215463713,
  "2": 1276327882,
  "3": 3565864927,
  "4": 2378129814,
  "5": 3101758893,
  "6": 2801591300,
  "7": 476782113,
  "8": 591441816,
  "9": 827183327,
  "10": 1679203998,
  "11": 3767535285,
  "12": 1968811824
}
