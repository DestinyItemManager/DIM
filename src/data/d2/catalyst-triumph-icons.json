{
  "221857936": "/common/destiny2_content/icons/5bcea1f9a52fb82286a26dca952c9259.jpg",
  "267836607": "/common/destiny2_content/icons/4a37bd1fc1756b09f509396557937832.jpg",
  "495916021": "/common/destiny2_content/icons/46ebd791f692b682691416beea7ac4c5.jpg",
  "543481464": "/common/destiny2_content/icons/a0ca8e1fddf00d1e2fa16357cabbd7f4.jpg",
  "745021037": "/common/destiny2_content/icons/0e4aaccfad36ae9179375de5477bfd13.jpg",
  "773877428": "/common/destiny2_content/icons/b760b737519af909e26f21009d6a1487.jpg",
  "1257836884": "/common/destiny2_content/icons/7dcb41a8e98004a1af64df680784b24d.jpg",
  "1366508788": "/common/destiny2_content/icons/2164d8411d065abce0442fe597b06a21.jpg",
  "1497347070": "/common/destiny2_content/icons/66fbe54a24d5a2b9978c909eb15ab62d.jpg",
  "1576480714": "/common/destiny2_content/icons/c013e41cdb32779bc2322337614ea06b.jpg",
  "1640774328": "/common/destiny2_content/icons/27fdad9f07de0dc4b97bf8574e6e5242.jpg",
  "1962727403": "/common/destiny2_content/icons/28355e0adda864289165854e376f86ee.jpg",
  "2087362844": "/common/destiny2_content/icons/3b8175fb92ee774d15d25fed58dca619.jpg",
  "2136108437": "/common/destiny2_content/icons/19e7c9a3d838e3801e7147975d173cb5.jpg",
  "2245186555": "/common/destiny2_content/icons/379a2872be1dc01faf4cae20508fe8bf.jpg",
  "2369153435": "/common/destiny2_content/icons/d011f6e099dc412ca11a7c8c12c08a3c.jpg",
  "2490851406": "/common/destiny2_content/icons/68df199792e5e6ab67de1befba7e0595.jpg",
  "2493039085": "/common/destiny2_content/icons/b22e5ede50882e552212533397f2443b.jpg",
  "2562666046": "/common/destiny2_content/icons/773f96cb2f9f2149ec31874cafec7528.jpg",
  "2660224414": "/common/destiny2_content/icons/1d38c8d82f1bd1d22bf76314fe8764f9.jpg",
  "2747759652": "/common/destiny2_content/icons/a6d9a19ac7f14078210b91b59d3955bb.jpg",
  "3121210389": "/common/destiny2_content/icons/060d820c7307f8cc9bc303965092214e.jpg",
  "3248675632": "/common/destiny2_content/icons/d9373b83e799a2d4035d82526ac362f3.jpg",
  "3250123968": "/common/destiny2_content/icons/5106f4624da0a5ff26b97702547fc282.jpg",
  "3337586114": "/common/destiny2_content/icons/3b4c1694f52b85f9fc1c57cb963dc36c.jpg",
  "3337586115": "/common/destiny2_content/icons/1bd8b5b8a2287f37ba549f09f46bd905.jpg",
  "3631669465": "/common/destiny2_content/icons/45d19f57bcbb22fa70cedca91a3f987e.jpg",
  "3686797699": "/common/destiny2_content/icons/013403d109a9563ee881830044177193.jpg",
  "3715769785": "/common/destiny2_content/icons/e68a5dba35f4b96f7230e88666e416e6.jpg",
  "3783249575": "/common/destiny2_content/icons/8e54c8921ee2947dc378088725197f22.jpg",
  "3812775859": "/common/destiny2_content/icons/e444e97b9d2c368006272fc90eb1e022.jpg",
  "3885534828": "/common/destiny2_content/icons/3e864b06fb9266cac7b4ab89835d2e6f.jpg",
  "3904776877": "/common/destiny2_content/icons/562b80b58bccce105b7cdd81847f665c.jpg",
  "3985015094": "/common/destiny2_content/icons/fc8734628c47ba0add073aef57f8e860.jpg",
  "4181440763": "/common/destiny2_content/icons/92577b2999d98670c8a6a98a9646eff8.jpg",
  "4243279770": "/common/destiny2_content/icons/d3b1aeb1757f2011bc062befe58ab3dd.jpg"
}
