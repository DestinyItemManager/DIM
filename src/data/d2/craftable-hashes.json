[
  14194600,
  45643573,
  46524085,
  135029084,
  231031173,
  232928045,
  254636484,
  268260372,
  268260373,
  297296830,
  318443586,
  392008588,
  431721920,
  484515708,
  502356570,
  542203595,
  548958835,
  613334176,
  768621510,
  820890091,
  867154247,
  927567426,
  999767358,
  1168625549,
  1184309824,
  1248372789,
  1289796511,
  1298815317,
  1311684613,
  1321506184,
  1366394399,
  1392919471,
  1399109800,
  1471212226,
  1473821207,
  1478986057,
  1491665733,
  1509167284,
  1526296434,
  1572896086,
  1679868061,
  1720503118,
  1731355324,
  1751893422,
  1911060537,
  1937552980,
  1959650777,
  1994645182,
  2034215657,
  2097055732,
  2119346509,
  2149683300,
  2194955522,
  2218569744,
  2221264583,
  2263839058,
  2272041093,
  2302346155,
  2323544076,
  2490988246,
  2508948099,
  2531963421,
  2534546147,
  2535142413,
  2595497736,
  2607304614,
  2708806099,
  2721157927,
  2778013407,
  2827764482,
  2856514843,
  2972949637,
  2978226043,
  2990047042,
  3055790362,
  3103325054,
  3107853529,
  3118061005,
  3175851496,
  3228096719,
  3257091166,
  3257091167,
  3281285075,
  3366545721,
  3371413056,
  3371413057,
  3371413059,
  3428521585,
  3489657138,
  3635821806,
  3849444474,
  3849810018,
  3865728990,
  3886416794,
  3920310144,
  3969066556,
  4067556514,
  4096943616,
  4225322581,
  4230965989,
  4248569242
]
