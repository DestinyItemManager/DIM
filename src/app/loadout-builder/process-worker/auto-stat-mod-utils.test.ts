import { StatHashes } from 'app/../data/d2/generated-enums';
import { buildCacheV2 } from './auto-stat-mod-utils';

describe('test artifice', () => {
  test('the cache works', () => {
    const cache = buildCacheV2(5);
    console.log(JSON.stringify(cache.statCaches[StatHashes.Mobility]));
  });
});
