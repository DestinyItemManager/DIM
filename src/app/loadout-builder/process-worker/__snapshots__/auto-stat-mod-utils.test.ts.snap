// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`process-utils auto mod structure different ways of hitting target stats with 0 remaining general mods (using artifice mods: true) 1`] = `
{
  "1": 1,
  "10": 1,
  "11": 1,
  "12": 1,
  "13": 1,
  "14": 1,
  "15": 1,
  "2": 1,
  "3": 1,
  "4": 1,
  "5": 1,
  "6": 1,
  "7": 1,
  "8": 1,
  "9": 1,
}
`;

exports[`process-utils auto mod structure different ways of hitting target stats with 1 remaining general mods (using artifice mods: true) 1`] = `
{
  "1": 2,
  "10": 3,
  "11": 3,
  "12": 3,
  "13": 3,
  "14": 3,
  "15": 3,
  "16": 2,
  "17": 2,
  "18": 2,
  "19": 2,
  "2": 2,
  "20": 2,
  "21": 1,
  "22": 1,
  "23": 1,
  "24": 1,
  "25": 1,
  "3": 2,
  "4": 2,
  "5": 2,
  "6": 3,
  "7": 3,
  "8": 3,
  "9": 3,
}
`;

exports[`process-utils auto mod structure different ways of hitting target stats with 3 remaining general mods (using artifice mods: false) 1`] = `
{
  "1": 1,
  "10": 2,
  "11": 2,
  "12": 2,
  "13": 2,
  "14": 2,
  "15": 2,
  "16": 2,
  "17": 2,
  "18": 2,
  "19": 2,
  "2": 1,
  "20": 2,
  "21": 1,
  "22": 1,
  "23": 1,
  "24": 1,
  "25": 1,
  "26": 1,
  "27": 1,
  "28": 1,
  "29": 1,
  "3": 1,
  "30": 1,
  "4": 1,
  "5": 1,
  "6": 2,
  "7": 2,
  "8": 2,
  "9": 2,
}
`;

exports[`process-utils auto mod structure different ways of hitting target stats with 5 remaining general mods (using artifice mods: false) 1`] = `
{
  "1": 1,
  "10": 2,
  "11": 2,
  "12": 2,
  "13": 2,
  "14": 2,
  "15": 2,
  "16": 3,
  "17": 3,
  "18": 3,
  "19": 3,
  "2": 1,
  "20": 3,
  "21": 3,
  "22": 3,
  "23": 3,
  "24": 3,
  "25": 3,
  "26": 3,
  "27": 3,
  "28": 3,
  "29": 3,
  "3": 1,
  "30": 3,
  "31": 2,
  "32": 2,
  "33": 2,
  "34": 2,
  "35": 2,
  "36": 2,
  "37": 2,
  "38": 2,
  "39": 2,
  "4": 1,
  "40": 2,
  "41": 1,
  "42": 1,
  "43": 1,
  "44": 1,
  "45": 1,
  "46": 1,
  "47": 1,
  "48": 1,
  "49": 1,
  "5": 1,
  "50": 1,
  "6": 2,
  "7": 2,
  "8": 2,
  "9": 2,
}
`;

exports[`process-utils auto mod structure different ways of hitting target stats with 5 remaining general mods (using artifice mods: true) 1`] = `
{
  "1": 2,
  "10": 4,
  "11": 6,
  "12": 6,
  "13": 6,
  "14": 6,
  "15": 6,
  "16": 8,
  "17": 8,
  "18": 8,
  "19": 8,
  "2": 2,
  "20": 8,
  "21": 10,
  "22": 10,
  "23": 10,
  "24": 10,
  "25": 10,
  "26": 11,
  "27": 11,
  "28": 11,
  "29": 11,
  "3": 2,
  "30": 11,
  "31": 11,
  "32": 11,
  "33": 11,
  "34": 11,
  "35": 11,
  "36": 10,
  "37": 10,
  "38": 10,
  "39": 10,
  "4": 2,
  "40": 10,
  "41": 8,
  "42": 8,
  "43": 8,
  "44": 8,
  "45": 8,
  "46": 6,
  "47": 6,
  "48": 6,
  "49": 6,
  "5": 2,
  "50": 6,
  "51": 4,
  "52": 4,
  "53": 4,
  "54": 4,
  "55": 4,
  "56": 2,
  "57": 2,
  "58": 2,
  "59": 2,
  "6": 4,
  "60": 2,
  "61": 1,
  "62": 1,
  "63": 1,
  "64": 1,
  "65": 1,
  "7": 4,
  "8": 4,
  "9": 4,
}
`;

exports[`process-utils auto mod structure snapshot of mod defs when assuming cheapgeneral for auto mods 1`] = `
{
  "artificeMods": {
    "144602215": {
      "hash": 3160845295,
    },
    "1735777505": {
      "hash": 617569843,
    },
    "1943323491": {
      "hash": 539459624,
    },
    "2996146975": {
      "hash": 2322202118,
    },
    "392767087": {
      "hash": 199176566,
    },
    "4244567218": {
      "hash": 2507624050,
    },
  },
  "generalMods": {
    "144602215": {
      "majorMod": {
        "cost": 4,
        "hash": 2724608735,
      },
      "minorMod": {
        "cost": 2,
        "hash": 350061697,
      },
    },
    "1735777505": {
      "majorMod": {
        "cost": 3,
        "hash": 1435557120,
      },
      "minorMod": {
        "cost": 1,
        "hash": 4021790309,
      },
    },
    "1943323491": {
      "majorMod": {
        "cost": 4,
        "hash": 4204488676,
      },
      "minorMod": {
        "cost": 2,
        "hash": 1237786518,
      },
    },
    "2996146975": {
      "majorMod": {
        "cost": 3,
        "hash": 4183296050,
      },
      "minorMod": {
        "cost": 1,
        "hash": 1703647492,
      },
    },
    "392767087": {
      "majorMod": {
        "cost": 4,
        "hash": 1180408010,
      },
      "minorMod": {
        "cost": 2,
        "hash": 2532323436,
      },
    },
    "4244567218": {
      "majorMod": {
        "cost": 3,
        "hash": 4287799666,
      },
      "minorMod": {
        "cost": 1,
        "hash": 2639422088,
      },
    },
  },
}
`;

exports[`process-utils auto mod structure snapshot of mod defs when assuming cheapgeneral for auto mods 2`] = `
{
  "0": [
    0,
    3,
    5,
  ],
  "1": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "2": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "3": [
    0,
    3,
    5,
  ],
  "4": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "5": [
    0,
    3,
    5,
  ],
}
`;

exports[`process-utils auto mod structure snapshot of mod defs when assuming general for auto mods 1`] = `
{
  "artificeMods": {
    "144602215": {
      "hash": 3160845295,
    },
    "1735777505": {
      "hash": 617569843,
    },
    "1943323491": {
      "hash": 539459624,
    },
    "2996146975": {
      "hash": 2322202118,
    },
    "392767087": {
      "hash": 199176566,
    },
    "4244567218": {
      "hash": 2507624050,
    },
  },
  "generalMods": {
    "144602215": {
      "majorMod": {
        "cost": 4,
        "hash": 2724608735,
      },
      "minorMod": {
        "cost": 2,
        "hash": 350061697,
      },
    },
    "1735777505": {
      "majorMod": {
        "cost": 3,
        "hash": 1435557120,
      },
      "minorMod": {
        "cost": 1,
        "hash": 4021790309,
      },
    },
    "1943323491": {
      "majorMod": {
        "cost": 4,
        "hash": 4204488676,
      },
      "minorMod": {
        "cost": 2,
        "hash": 1237786518,
      },
    },
    "2996146975": {
      "majorMod": {
        "cost": 3,
        "hash": 4183296050,
      },
      "minorMod": {
        "cost": 1,
        "hash": 1703647492,
      },
    },
    "392767087": {
      "majorMod": {
        "cost": 4,
        "hash": 1180408010,
      },
      "minorMod": {
        "cost": 2,
        "hash": 2532323436,
      },
    },
    "4244567218": {
      "majorMod": {
        "cost": 3,
        "hash": 4287799666,
      },
      "minorMod": {
        "cost": 1,
        "hash": 2639422088,
      },
    },
  },
}
`;

exports[`process-utils auto mod structure snapshot of mod defs when assuming general for auto mods 2`] = `
{
  "0": [
    0,
    3,
    5,
  ],
  "1": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "2": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "3": [
    0,
    3,
    5,
  ],
  "4": [
    0,
    1,
    2,
    3,
    4,
    5,
  ],
  "5": [
    0,
    3,
    5,
  ],
}
`;
